### DISCUSSION TITLE
MapStruct did not autowired other specified mappers

### DISCUSSION THREAD

**Discussion by TenPetr** [mapstruct/mapstruct#2576] [Q&A]
I am trying to map entity fields to the DTO in my code - Property `storage` can be either `Location` or `Container` (in GraphQL we used a union for it), but there is a problem with mapping using mapstruct. In the database, it is saved to two nullable fields - `storage_container_id` and `storage_location_id`, and one of these fields can be null. 

I implemented a custom mapper, which checked if one is null and if so, it would use a second property. (if the container is null, then use location). In this custom mapper, I called other mappers that would map either Location entity to DAO or Container entity to DAO, but I cannot use it in my class, because they are not autowired in the constructor. I tried multiple approaches, but non of them worked.

- I tried `expression = java( "logic here" )`

- Use `abstract class` and autowire mappers

```
error: constructor ItemMapper in class ItemMapper cannot be applied to given types;
    public ItemMapperImpl(ArticleMapper articleMapper) {
                                                       ^
  required: ResourceMapper,LocationMapper
```

- `@Context` annotation with two mappers (this approach generated weird recursive code that did not call these services at all)

Do you have any idea, why is this happening? Thanks a lot.

**Comment by filiphr** [comment]
@TenPetr please do not use the issue tracker for asking questions.

MapStruct does not autowire other mappers / services from `Mapper#uses` which are not explicitly used. In order for this to work you need to do the auto wiring on your own side in an abstract class. In addition to that you'll need to use setter injection, otherwise you will have the issue you are having.
